deepseg.cc

```diff
@@ -244,59 +289,61 @@ void calc_mask(calcinfo_t &info, timinginfo_t &ti) {
-	if (strstr(info.modelname,"deeplab")) {
-		for (unsigned int n = 0; n < info.output.total(); n++) {
-			float maxval = -10000; size_t maxpos = 0;
-			for (size_t i = 0; i < cnum; i++) {
-				if (tmp[n*cnum+i] > maxval) {
-					maxval = tmp[n*cnum+i];
-					maxpos = i;
+	switch (info.modeltype) {
+		case modeltype_t::DeepLab:
+			// find class with maximum probability
+			for (unsigned int n = 0; n < info.output.total(); n++) {
+				float maxval = -10000; size_t maxpos = 0;
+				for (size_t i = 0; i < cnum; i++) {
+					if (tmp[n*cnum+i] > maxval) {
+						maxval = tmp[n*cnum+i];
+						maxpos = i;
+					}
 				}
+				// set mask to 0 where class == person
+				uint8_t val = (maxpos==pers ? 0 : 255);
+				out[n] = (val & 0xE0) | (out[n] >> 3);
 			}
-			// set mask to 0 where class == person
-			uint8_t val = (maxpos==pers ? 0 : 255);
-			out[n] = (val & 0xE0) | (out[n] >> 3);
-		}
-	}
-
-	// threshold probability
-	if (strstr(info.modelname,"body-pix") || strstr(info.modelname,"selfie")) {
-		for (unsigned int n = 0; n < info.output.total(); n++) {
-			// FIXME: hardcoded threshold
-			uint8_t val = (tmp[n] > 0.65 ? 0 : 255);
-			out[n] = (val & 0xE0) | (out[n] >> 3);
-		}
-	}
-
-	// Google Meet segmentation network
-	if (strstr(info.modelname,"segm_")) {
-		/* 256 x 144 x 2 tensor for the full model or 160 x 96 x 2
-		 * tensor for the light model with masks for background
-		 * (channel 0) and person (channel 1) where values are in
-		 * range [MIN_FLOAT, MAX_FLOAT] and user has to apply
-		 * softmax across both channels to yield foreground
-		 * probability in [0.0, 1.0]. */
-		for (unsigned int n = 0; n < info.output.total(); n++) {
-			float exp0 = expf(tmp[2*n  ]);
-			float exp1 = expf(tmp[2*n+1]);
-			float p0 = exp0 / (exp0+exp1);
-			float p1 = exp1 / (exp0+exp1);
-			uint8_t val = (p0 < p1 ? 0 : 255);
-			out[n] = (val & 0xE0) | (out[n] >> 3);
-		}
+			break;
+		case modeltype_t::BodyPix:
+		case modeltype_t::MLKitSelfie:
+			// threshold probability
+			for (unsigned int n = 0; n < info.output.total(); n++) {
+				// FIXME: hardcoded threshold
+				uint8_t val = (tmp[n] > 0.65 ? 0 : 255);
+				out[n] = (val & 0xE0) | (out[n] >> 3);
+			}
+			break;
+		case modeltype_t::GoogleMeetSegmentation:
+			/* 256 x 144 x 2 tensor for the full model or 160 x 96 x 2
+			 * tensor for the light model with masks for background
+			 * (channel 0) and person (channel 1) where values are in
+			 * range [MIN_FLOAT, MAX_FLOAT] and user has to apply
+			 * softmax across both channels to yield foreground
+			 * probability in [0.0, 1.0]. */
+			for (unsigned int n = 0; n < info.output.total(); n++) {
+				float exp0 = expf(tmp[2*n  ]);
+				float exp1 = expf(tmp[2*n+1]);
+				float p0 = exp0 / (exp0+exp1);
+				float p1 = exp1 / (exp0+exp1);
+				uint8_t val = (p0 < p1 ? 0 : 255);
+				out[n] = (val & 0xE0) | (out[n] >> 3);
+			}
+			break;
+		case modeltype_t::Unknown:
+			fprintf(stderr, "Unknown model type\n");
```

Comment:
What about assuming direct mask as output in this case?

Reply:
I.e. use same as BodyPix?

Reply:
@vekkt0r did you want to make this change before I merge the PR? I believe this is the last remaining point.